### ISSUE TITLE
Google API permissions error when using GKE runners

### ISSUE THREAD

**Issue by mihma** [google-github-actions/upload-cloud-storage#343]
### TL;DR

I have GKE runners with workload identity configured. I have confirmed the runner pods are using the correct Google service account. That SA has Object Admin role assigned to a specific bucket. When I run the upload action it fails with this error:
`GaxiosError: Could not refresh access token: Unsuccessful response status code. Request failed with status code 403
...
data: "Unable to generate access token; IAM returned 403 Forbidden: Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).\n"
...
scopes: 'https://www.googleapis.com/auth/iam,https://www.googleapis.com/auth/cloud-platform,https://www.googleapis.com/auth/devstorage.full_control'`

The same works fine if I use auth action beforehand with JSON credentials supplied. It seems as if when using [Application Default Credentials](https://github.com/google-github-actions/upload-cloud-storage/issues/new?assignees=&labels=bug&projects=&template=bug.yaml#via-application-default-credentials) it requires to have much wider permissions: `https://www.googleapis.com/auth/devstorage.full_control`?

### Action YAML

```yaml
- uses: 'google-github-actions/upload-cloud-storage@v1'
  with:
    path: './public'
    destination: 'bucket.name'
```

### Log output

```text
Uploading /runner/_work/name.xlsx to gs://bucket.name/public/data-samples/name.xlsx
29
  Error: google-github-actions/upload-cloud-storage failed with: could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: A Forbidden error was returned while attempting to retrieve an access token for the Compute Engine built-in service account. This may be because the Compute Engine instance does not have the correct permission scopes specified: Could not refresh access token: Unsuccessful response status code. Request failed with status code 403
30
  node:internal/process/promises:279
31
              triggerUncaughtException(err, true /* fromPromise */);
32
              ^
33
  
34
  GaxiosError: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: Could not refresh access token: A Forbidden error was returned while attempting to retrieve an access token for the Compute Engine built-in service account. This may be because the Compute Engine instance does not have the correct permission scopes specified: Could not refresh access token: Unsuccessful response status code. Request failed with status code 403
35
      at Gaxios._request (/runner/_work/_actions/google-github-actions/upload-cloud-storage/v1/dist/index.js:183:5830)
36
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
37
      at async metadataAccessor (/runner/_work/_actions/google-github-actions/upload-cloud-storage/v1/dist/index.js:183:13775)
38
      at async Compute.refreshTokenNoCache (/runner/_work/_actions/google-github-actions/upload-cloud-storage/v1/dist/index.js:183:30037) ***
39
    response: ***
40
      config: ***
41
        url: 'http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token?scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fiam%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdevstorage.full_control',
42
        headers: *** 'Metadata-Flavor': 'Google' ***,
43
        retryConfig: ***
44
          noResponseRetries: 3,
45
          currentRetryAttempt: 0,
46
          retry: 3,
47
          httpMethodsToRetry: [ 'GET', 'HEAD', 'PUT', 'OPTIONS', 'DELETE' ],
48
          statusCodesToRetry: [ [ 100, 199 ], [ 429, 429 ], [ 500, 599 ] ]
49
        ***,
50
        params: ***
51
          scopes: 'https://www.googleapis.com/auth/iam,https://www.googleapis.com/auth/cloud-platform,https://www.googleapis.com/auth/devstorage.full_control'
52
        ***,
53
        responseType: 'text',
54
        timeout: 0,
55
        paramsSerializer: [Function: paramsSerializer],
56
        validateStatus: [Function: validateStatus],
57
        method: 'GET'
58
      ***,
59
      data: "Unable to generate access token; IAM returned 403 Forbidden: Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).\n" +
60
        'This error could be caused by a missing IAM policy binding on the target IAM service account.\n' +
61
        'For more information, refer to the Workload Identity documentation:\n' +
62
        '\thttps://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#authenticating_to\n' +
63
        '\n',
64
      headers: ***
65
        connection: 'close',
66
        'content-length': '406',
67
        'content-type': 'text/plain; charset=utf-8',
68
        date: 'Thu, 26 Oct 2023 14:23:59 GMT',
69
        'x-content-type-options': 'nosniff'
70
      ***,
71
      status: 403,
72
      statusText: 'Forbidden',
73
      request: ***
74
        responseURL: 'http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token?scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fiam%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdevstorage.full_control'
75
      ***
76
    ***,
77
    config: ***
78
      url: 'http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token?scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fiam%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdevstorage.full_control',
79
      headers: *** 'Metadata-Flavor': 'Google' ***,
80
      retryConfig: ***
81
        noResponseRetries: 3,
82
        currentRetryAttempt: 0,
83
        retry: 3,
84
        httpMethodsToRetry: [ 'GET', 'HEAD', 'PUT', 'OPTIONS', 'DELETE' ],
85
        statusCodesToRetry: [ [ 100, 199 ], [ 429, 429 ], [ 500, 599 ] ]
86
      ***,
87
      params: ***
88
        scopes: 'https://www.googleapis.com/auth/iam,https://www.googleapis.com/auth/cloud-platform,https://www.googleapis.com/auth/devstorage.full_control'
89
      ***,
90
      responseType: 'text',
91
      timeout: 0,
92
      paramsSerializer: [Function: paramsSerializer],
93
      validateStatus: [Function: validateStatus],
94
      method: 'GET'
95
    ***,
96
    code: '403'
```

**Comment by sethvargo** [comment]
Hi @mihma - that error is coming directly from the Google Cloud Storage Node.js library. What steps did you take to verify that the Workload Identity is configured and working correctly?

**Comment by mihma** [comment]
Hi @sethvargo, thanks for getting back to me.
I'm using those steps to configure the workload identity: https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity
And verify it by creating a pod with the same k8s service account I use for runner pods as described here https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#verify_the_setup
It gives me the correct GCP service account email address.

**Comment by sethvargo** [comment]
Thanks @mihma - can you please try following those exact same steps, but running the following `curl` command at the end instead and reporting the result:

```shell
curl -sv -H "Metadata-Flavor: Google" "http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token?scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fiam%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdevstorage.full_control"
```

**Comment by mihma** [comment]
`root@workload-identity-test:/# curl -sv -H "Metadata-Flavor: Google" "http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token?scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fiam%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdevstorage.full_control"
*   Trying 169.254.169.254:80...
* Connected to 169.254.169.254 (169.254.169.254) port 80 (#0)
> GET /computeMetadata/v1/instance/service-accounts/default/token?scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fiam%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdevstorage.full_control HTTP/1.1
> Host: 169.254.169.254
> User-Agent: curl/7.74.0
> Accept: */*
> Metadata-Flavor: Google
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 403 Forbidden
< Content-Type: text/plain; charset=utf-8
< X-Content-Type-Options: nosniff
< Date: Fri, 27 Oct 2023 15:46:31 GMT
< Content-Length: 406
< 
Unable to generate access token; IAM returned 403 Forbidden: Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).
This error could be caused by a missing IAM policy binding on the target IAM service account.
For more information, refer to the Workload Identity documentation:
	https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#authenticating_to

* Connection #0 to host 169.254.169.254 left intact`

I have just tried and created the same pod with the default k8s sa in another namespace and your command was able to get the token..  I will spend some more time troubleshooting this and will report back here.

**Comment by sethvargo** [comment]
Hi @mihma - that indicates that the service account does not have permission to generate tokens.

**Comment by mihma** [comment]
I've just recreated google service account and is now able to get a token. The upload action also works smoothly! 
I will assume that was some kind of service account glitch from GCP side. Thanks for your help, I appreciate it!
